drivers/gpio/gpio_gecko.c

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright (c) 2017, Christian Taedcke
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <gpio.h>
+#include <soc.h>
+#include <em_gpio.h>
+
+#include "gpio_utils.h"
+
+#define member_size(type, member) sizeof(((type *)0)->member)
+#define NUMBER_OF_PORTS (member_size(GPIO_TypeDef, P) / \
+			 member_size(GPIO_TypeDef, P[0]))
+
+struct gpio_gecko_common_config {
+};
+
+struct gpio_gecko_common_data {
+	/* a list of all ports */
+	struct device *ports[NUMBER_OF_PORTS];
+	size_t count;
+};
+
+struct gpio_gecko_config {
+	GPIO_P_TypeDef *gpio_base;
+	GPIO_Port_TypeDef gpio_index;
+};
+
+struct gpio_gecko_data {
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+	/* pin callback routine enable flags, by pin number */
+	u32_t pin_callback_enables;
+};
+
+static inline void gpio_gecko_add_port(struct gpio_gecko_common_data *data,
+				       struct device *dev)
+{
+	__ASSERT(dev, "No port device!");
+	data->ports[data->count++] = dev;
+}
+
+
+static int gpio_gecko_configure(struct device *dev,
+				int access_op, u32_t pin, int flags)
+{
+	const struct gpio_gecko_config *config = dev->config->config_info;
+	GPIO_P_TypeDef *gpio_base = config->gpio_base;
+	GPIO_Port_TypeDef gpio_index = config->gpio_index;
+	GPIO_Mode_TypeDef mode;
+	unsigned int out = 0;
+	bool interrupt_rising_edge = false;
+	bool interrupt_falling_edge = false;
+	bool enable_interrupt = false;
+
+	/* Check for an invalid pin configuration */
+	if ((flags & GPIO_INT) && (flags & GPIO_DIR_OUT)) {
+		return -EINVAL;
+	}
+
+	/* Interrupt on static level is not supported by the hardware */
+	if ((flags & GPIO_INT) && ((flags & GPIO_INT_EDGE) == GPIO_INT_LEVEL)) {
```

Comment:
the second test: write !(flags & GPIO_INT_EDGE)
testing for EDGE that equals LEVEL is weird to read (I know, it's because LEVEL is 0, and EDGE 1).